### PR TITLE
App layer stats v17

### PR DESCRIPTION
**Pull request by regit**
Update of https://github.com/inliniac/suricata/pull/2116 addressing comments and rebasing on latest master.

PR builds:
- PR regit: https://buildbot.openinfosecfoundation.org/builders/regit/builds/188
- PR regit-pcap: https://buildbot.openinfosecfoundation.org/builders/regit-pcap/builds/184

### CODE REVIEW

File: src/app-layer.c [hunk 1]
```diff
@@ -648,6 +678,80 @@ void AppLayerRegisterGlobalCounters(void)
...
+                }
+                snprintf(applayer_counters[ipproto_map][alproto].name,
+                         strlen(str) + alproto_len,
+                         "%s%s", str, alproto_str);
+            }
+        }
+    }
+}
+
+void AppLayerRegisterCounters(ThreadVars *tv, uint8_t ipproto)
```
Comment (inliniac) [inline src/app-layer.c]: This is per thread so that should be reflected in the name.

File: src/app-layer.c [hunk 2]
```diff
@@ -648,6 +678,80 @@ void AppLayerRegisterGlobalCounters(void)
...
+    AppLayerProtoDetectSupportedAppProtocols(alprotos);
+
+    for (alproto = 0; alproto < ALPROTO_MAX; alproto++) {
+        if (alprotos[alproto] == 1) {
+            char *str = "app_layer.flow.";
+            char *alproto_str = AppLayerGetProtoName(alproto);
+            int alproto_len = strlen(alproto_str) + 1;
+            uint8_t ipproto_map = FlowGetProtoMapping(ipproto);
+
+            if (AppLayerParserProtoIsRegistered(ipproto, alproto) &&
```
Comment (inliniac) [inline src/app-layer.c]: Can you explain the logic here? Don't see what this check does and why there are 2 conditions.

File: src/app-layer.c [hunk 3]
```diff
@@ -648,6 +678,80 @@ void AppLayerRegisterGlobalCounters(void)
...
+    for (alproto = 0; alproto < ALPROTO_MAX; alproto++) {
+        if (alprotos[alproto] == 1) {
+            uint8_t ipproto_map = FlowGetProtoMapping(ipproto);
+            applayer_counters[ipproto_map][alproto].counter_id =
+                StatsRegisterCounter(applayer_counters[ipproto_map][alproto].name, tv);
+        }
+    }
+}
+
+void AppLayerDeRegisterCounters(uint8_t ipproto)
```
Comment (inliniac) [inline src/app-layer.c]: This is not per thread, but it does have a name 'opposite' of a thread func: AppLayerRegisterCounters (per thread) vs AppLayerDeRegisterCounters (global)

File: src/flow-worker.c [hunk 4]
```diff
@@ -98,6 +98,9 @@ static TmEcode FlowWorkerThreadInit(ThreadVars *tv, void *initdata, void **data)
         return TM_ECODE_FAILED;
     }
 
+    AppLayerRegisterCounters(tv, IPPROTO_UDP);
```
Comment (inliniac) [inline src/flow-worker.c]: Not sure why we should have 2 calls here for TCP and UDP. Might just as well move the ipproto inside the Register func.

File: src/output-json-stats.c [hunk 5]
```diff
@@ -47,6 +47,8 @@
 #include "output-json.h"
 #include "output-json-stats.h"
 
+#include "app-layer.h"
```
Comment (inliniac) [inline src/output-json-stats.c]: Seems unnecessary in this commit?

File: src/app-layer.c [hunk 6]
```diff
@@ -576,6 +600,9 @@ int AppLayerSetup(void)
     AppLayerParserRegisterProtocolParsers();
     AppLayerProtoDetectPrepareState();
 
+    AppLayerSetupCounters(IPPROTO_UDP);
```
Comment (inliniac) [inline src/app-layer.c]: Same here, we can move the ipproto logic inside the Setup func

File: src/app-layer.c [hunk 7]
```diff
@@ -586,6 +613,9 @@ int AppLayerDeSetup(void)
     AppLayerProtoDetectDeSetup();
     AppLayerParserDeSetup();
 
+    AppLayerDeRegisterCounters(IPPROTO_UDP);
```
Comment (inliniac) [inline src/app-layer.c]: Same here, we can move the ipproto logic inside the Setup func

File: src/app-layer.c [hunk 8]
```diff
@@ -80,6 +91,14 @@ static inline int ProtoDetectDone(const Flow *f, const TcpSession *ssn, uint8_t
             (FLOW_IS_PM_DONE(f, direction) && FLOW_IS_PP_DONE(f, direction)));
 }
 
+static void AppLayerIncFlowCounter(ThreadVars *tv, uint8_t ipproto, AppProto alproto)
```
Comment (inliniac) [inline src/app-layer.c]: If you just pass 'Flow *f' to this function you can directly use f->protomap and f->alproto. Safes one FlowGetProtoMapping call.

File: src/app-layer-parser.c [hunk 9]
```diff
@@ -1011,6 +1014,20 @@ int AppLayerParserParse(ThreadVars *tv, AppLayerParserThreadCtx *alp_tctx, Flow
...
     }
 
+    if (AppLayerParserProtocolIsTxAware(f->proto, alproto)) {
+        if (likely(tv)) {
+            /* At each pass in the parser we add to the tv counter
+             * the difference of tx on a flow between now and the previous pass
+             */
+            uint64_t p_tx_cnt = pstate->tx_cnt;
+            pstate->tx_cnt = AppLayerParserGetTxCnt(f->proto, alproto, f->alstate);
+            if (!AppLayerParserProtocolIsTxAware(f->proto, alproto)) {
```
Comment (inliniac) [inline src/app-layer-parser.c]: Seems already checked on line 1017?

File: src/app-layer-parser.c [hunk 10]
```diff
@@ -140,6 +140,8 @@ struct AppLayerParserState_ {
      * we don't need a var per direction since we don't log a transaction
      * unless we have the entire transaction. */
     uint64_t log_id;
+    /* Count of number of transactions */
+    uint64_t tx_cnt;
```
Comment (inliniac) [inline src/app-layer-parser.c]: I'm not convinced this is needed if it's only used to get a diff between tx_cnt before and after a AppLayerParse call.

File: src/app-layer.c [hunk 11]
```diff
@@ -99,6 +101,15 @@ static void AppLayerIncFlowCounter(ThreadVars *tv, uint8_t ipproto, AppProto alp
     }
 }
 
+void AppLayerIncTxCounter(ThreadVars *tv, uint8_t ipproto, AppProto alproto,
+                          uint64_t step)
+{
+    if (likely(tv)) {
+        uint8_t ipproto_map = FlowGetProtoMapping(ipproto);
```
Comment (inliniac) [inline src/app-layer.c]: We can avoid this if the Flow is passed to this function. Then we can use f->protomap

File: src/app-layer-parser.c [hunk 12]
```diff
@@ -1011,6 +1014,20 @@ int AppLayerParserParse(ThreadVars *tv, AppLayerParserThreadCtx *alp_tctx, Flow
         }
     }
 
+    if (AppLayerParserProtocolIsTxAware(f->proto, alproto)) {
+        if (likely(tv)) {
+            /* At each pass in the parser we add to the tv counter
```
Comment (regit) [inline src/app-layer-parser.c]: @inliniac on the tx_cnt being part of the Flow structure, we have tried to explain it in this comment.

Comment (inliniac) [inline src/app-layer-parser.c]: I don't see why you need it to be in the state.

Why would this not work?
<pre>
p_tx_cnt = ...GetTxCnt();
...
Parser Runs
...
cur_tx_cnt = ...GetTxCnt()
if (cur_tx_cnt > p_tx_cnt) {
    new_tx_cnt = cur_tx_cnt - p_tx_cnt;
}
</pre>

File: src/app-layer.c (hunk 2 above)
Comment (glongo) [inline src/app-layer.c]: Here we check if a protocol, like dns, is registered for both tcp and udp. if yes, we add a suffix (ipproto_suffix) to the counter name, for instance: "dns_tcp", or "dns_udp".

**Pull request closed without merging**